driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,14 +337,33 @@ private void dequeue(Connection connection) {
                 }
             }
 
-            SettableFuture<Connection> pendingBorrow = pendingBorrows.poll();
+            final SettableFuture<Connection> pendingBorrow = pendingBorrows.poll();
             if (pendingBorrow == null) {
                 // Another thread has emptied the queue since our last check, restore the count
                 connection.inFlight.decrementAndGet();
             } else {
                 totalInFlight.incrementAndGet();
                 pendingBorrowCount.decrementAndGet();
-                pendingBorrow.set(connection);
+                // Ensure that the keyspace set on the connection is the one set on the pool state, in the general case it will be.
+                ListenableFuture<Connection> setKeyspaceFuture = connection.setKeyspaceAsync(manager.poolsState.keyspace);
+                // Slight optimization, if the keyspace was already correct the future will be complete, so simply complete it here.
+                if (setKeyspaceFuture.isDone()) {
```

Comment:
There's a bug here.  If the future is done and failed, we don't mark this future as failed and we probably should.  Also, if the future fails, we should probably try again instead of remaining in a broken state.  We should also be reactive to this too in `setKeyspaceAsync`